new-log-viewer/src/components/Layout.tsx

```diff
@@ -133,30 +148,97 @@ const ConfigForm = () => {
     );
 };
 
+/**
+ * Handles `logEventNum` input value change for debugging.
+ *
+ * @param ev
+ */
+const handleLogEventNumInputChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
+    updateWindowUrlHashParams({logEventNum: Number(ev.target.value)});
+};
+
 /**
  * Renders the major layout of the log viewer.
  *
  * @return
  */
 const Layout = () => {
     const {
-        logData,
         pageNum,
         numEvents,
     } = useContext(StateContext);
     const {logEventNum} = useContext(UrlContext);
 
-    const handleLogEventNumInputChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
-        updateWindowUrlHashParams({logEventNum: Number(ev.target.value)});
-    };
+    const logEventNumRef = useRef<Nullable<number>>(logEventNum);
+    const numEventsRef = useRef<Nullable<number>>(numEvents);
 
     const handleCopyLinkButtonClick = () => {
         copyPermalinkToClipboard({}, {logEventNum: numEvents});
     };
 
+    /**
+     * Handles custom actions in the editor.
+     *
+     * @param editor
+     * @param actionName The custom action to perform.
+     */
+    const handleCustomAction = useCallback((
+        editor: monaco.editor.IStandaloneCodeEditor,
+        actionName: ACTION_NAME
+    ) => {
+        const pageSize = getConfig(CONFIG_KEY.PAGE_SIZE);
+
+        switch (actionName) {
+            case ACTION_NAME.FIRST_PAGE:
+                updateWindowUrlHashParams({logEventNum: 1});
+                break;
+            case ACTION_NAME.PREV_PAGE:
+                if (null !== logEventNumRef.current) {
+                    updateWindowUrlHashParams({
+                        logEventNum: (logEventNumRef.current <= pageSize) ?
+                            1 :
+                            getLastItemNumInPrevChunk(logEventNumRef.current, pageSize),
```

Comment:
I think `getLastItemNumInPrevChunk` is already clamped to a min of 1, so do we need this?

Reply:
`getLastItemNumInPrevChunk` doesn't clamp the value to 1. In fact, I think it makes sense to move the ternary assignment into `getLastItemNumInPrevChunk`. done in the next commit

Reply:
Oh right, missed that we multiply getChunkNum * chunkSize.